rust/src/write.rs

```diff
@@ -175,20 +176,35 @@ impl WriteOptions {
     }
 
     /// Creates a [`Writer`] whch writes to `w` using the given options
-    pub fn create<'a, W: Write + Seek>(self, w: W) -> McapResult<Writer<'a, W>> {
+    pub fn create<W: Write + Seek>(self, w: W) -> McapResult<Writer<W>> {
         Writer::with_options(w, self)
     }
 }
 
+#[derive(Hash, PartialEq, Eq)]
+struct ChannelContent<'a> {
+    topic: Cow<'a, str>,
+    schema_id: u16,
+    message_encoding: Cow<'a, str>,
+    metadata: Cow<'a, BTreeMap<String, String>>,
+}
+
+#[derive(Hash, PartialEq, Eq)]
+struct SchemaContent<'a> {
+    name: Cow<'a, str>,
```

Comment:
Could you ELI5 why we need Cow here?

Reply:
i'll add a comment. We use CoW because we want to compare equality between all of these values without neccessarily cloning or moving all of them into a struct to do that.

Reply:
Is there some overhead inherent in the Cow struct itself? Should I think of as being basically a union between a couple different storage options/types? or a pointer to an allocation elsewhere?

Reply:
a Cow is a union between a borrowed and owned version of a given type - so for str, it's (String | &str), for bytes it's (&[u8], Vec<u8>). It implements Deref for those types, allowing you to use their comparison, hash implementations against eachother.

  In context, this struct lets me solves this problem:
```
struct left = {
  a: String::new("abc"),
  b: String::new("def")
}

struct right = {
  a: "abc",
  b: "def",
}
if a == b {
   println!("they are the same!")
}
```